### ISSUE TITLE
t/io/socket.t fails on OmniOS using GCC but not G++

### ISSUE THREAD

**Issue by cjg-cguevara** [Perl/perl5#20188]
One example:
https://perl5.test-smoke.org/report/5022137
https://perl5.test-smoke.org/logfile/5022137

https://perl5.test-smoke.org/report/5022146
https://perl5.test-smoke.org/logfile/5022146

**Comment by jkeenan** [comment]
https://perl5.test-smoke.org/submatrix?test=../t/porting/authors.t suggests that this emerged between 5.37.1 and 5.37.2 -- but was that just an artifact of when you first tested on this platform?  If not, can you bisect?

**Comment by cjg-cguevara** [comment]
I don't have access to my host cjg-omniosce right now, but it looks like it started failing sometime after 431ea43db266 on hosts perlsmoke.srv.customer.skylime.net and cjg-omniosce.

**Comment by bram-perl** [comment]
I've been somewhat looking somewhat at t/io/socket.t, test 45 that fails was added in commit f192c227eefb2861335e0592230df7f6d7607abf

What the test does:
- it uses `setsockopt` to change the size of the receive buffer
- it then uses `getsockopt` to query the size

The test succeeds if the value returned by `getsockopt` is:
- the same as the value used in the `setsockopt` call
- twice the size as the value used in the `setsockopt` call (see below as to why)

On your smoker:
- `setsockopt` was called with value 32768 
- `getsockopt` returned the value 33232  (= 32768 + 464)

Since `getsockopt` returned a value the test wasn't expecting the test failed.

___

The reason for the '2 * RCVBUF_SIZE' is because Linux doubles the buffer size.

Some linux related threads:
- https://lkml.iu.edu/hypermail/linux/kernel/1207.2/00685.html
- https://lkml.iu.edu/hypermail/linux/kernel/0208.3/0295.html
- https://www.cs.helsinki.fi/linux/linux-kernel/2001-30/0880.html

The last one is the most interesting one:

> You are making assumptions not guaranteed in POSIX or SuS. In the Linux case we deliberately allow more than requested as our memory accounting behaviour for buffers is very different to BSD

I tried to find what POSIX actually guarantees but so far I had no luck with that..
I **think** the test is too strict..
Doing some more testing on that smoker (/on Solaris) could be useful..
Does `getsockopt` always returns a value that is 464 bytes greater? does it differ? ..?
Unfortunately I don't have access to such a system so I can't test it myself...

**Comment by bram-perl** [comment]
Doing some testing on OmniOS, using following C program:

```
#include <stdio.h>
#include <sys/socket.h>

int main () {
        int sockfd = socket(AF_INET, SOCK_STREAM, 0);
        if (sockfd < 0) {
                fprintf(stderr, "Failed to create socket\n");
                return 1;
        }

        int orig;
        unsigned int len = sizeof(len);
        int res = getsockopt(sockfd, SOL_SOCKET, SO_RCVBUF, &orig, &len);
        if (res != 0) {
                fprintf(stderr, "initial getsockopt failed, res=%d\n", res);
                return 1;
        }
        printf("Initial value of SO_RCVBUF=%d\n", orig);


        for (int i = 1; i < 256000; i++) {
                int val;
                res = setsockopt(sockfd, SOL_SOCKET, SO_RCVBUF, &i, len);
                if (res != 0) {
                        fprintf(stderr, "setsockopt %d failed, res=%d\n", i, res);
                        return 1;
                }

                res = getsockopt(sockfd, SOL_SOCKET, SO_RCVBUF, &val, &len);
                if (res != 0) {
                        fprintf(stderr, "getsockopt (%d) failed, res=%d\n", i, res);
                        return 1;
                }
                printf("setsockopt(%d), getsockopt() returns %d, diff: %d\n", i, val, (val - i));
        }

        return 0;
}
```

Compiling and running:
```
$ gcc -lsocket gh-20188.c && ./a.out > o.gcc
$ g++ -lsocket gh-20188.c && ./a.out > o.g++
```

That shows:
- for g++ the difference is always 0. I.e. what was used in `setsockopt` is returned by `getsockopt`
- for gcc:

setsockopt value | getsockopt returns
--|--
1 | 2144
500 | 2144
2143 | 2144
2144 | 2144
2145 | 2680
2680 | 2680
2681 | 3216
3217 | 3752
... | ...
32696 | 32699
32768 | 33232
33232 | 33232
33233 | 33768

Basically:
- minimum value is 2144
- after that: `setsockopt($i)` --> `getsockopt()` returns: `ceil($i / 536) * 536`

So it is the test that is too picky..

I also have no idea why behavior of gcc and g++ differ nor what the significance of 536 is..

**Comment by jkeenan** [comment]
> I've been somewhat looking somewhat at t/io/socket.t, test 45 that fails was added in commit [f192c22](https://github.com/Perl/perl5/commit/f192c227eefb2861335e0592230df7f6d7607abf)
> 

Good research!

If you do have access to this platform ...

(a) can you bisect to determine the point where the Perl test failure first emerged?

(b) can you suggest a more appropriate unit test, and then test it out in a smoke-me branch? (@cjg-cguevara 's smoke-testing rigs are usually set up to automatically pick up such branches from the Perl5 repo.)

Thank you very much.
Jim Keenan

**Comment by bram-perl** [comment]
> (a) can you bisect to determine the point where the Perl test failure first emerged?

That is already in my earlier message.
The test started failing **the moment it was added**.
There is no point at all in doing a bisect.

**Comment by wolfsage** [comment]
SmartOS apparently rounds using the maximum segment size (by default 536 bytes):

https://github.com/TritonDataCenter/illumos-joyent/blob/master/usr/src/stand/lib/tcp/tcp.c#L7034

https://github.com/TritonDataCenter/illumos-joyent/blob/master/usr/src/stand/lib/tcp/tcp.c#L193

**Comment by bram-perl** [comment]
> SmartOS apparently rounds using the maximum segment size (by default 536 bytes):

And so is OmniOS.
Changing the value of tcp_mss: `ndd -set /dev/tcp tcp_mss_def_ipv4 100` and re-running shows:
- minimum value of 400 (i.e. 4 * 100)
- rounding to the next 100

Changing the value of tcp_recv_hiwat_minmss: `ndd -set /dev/tcp tcp_recv_hiwat_minmss 10` and re-running shows:
- minimum value of 1000 (i.e. 10 * 100)
- rounding to the next 100
